addons/payment/controllers/portal.py

```diff
@@ -1,376 +1,421 @@
+        :param dict kwargs: Locally unused data passed to `_is_tokenization_required` and
+                            `_compute_reference`
+        :return: The sudoed transaction that was created
+        :rtype: recordset of `payment.transaction`
+        :raise: UserError if the flow is invalid
+        """
+        # Prepare create values
+        if flow in ['redirect', 'direct']:  # Direct payment or payment with redirection
+            acquirer_sudo = request.env['payment.acquirer'].sudo().browse(payment_option_id)
+            token_id = None
+            tokenization_required_or_requested = acquirer_sudo._is_tokenization_required(
+                provider=acquirer_sudo.provider, **kwargs
+            ) or tokenization_requested
+            tokenize = bool(
+                # Public users are not allowed to save tokens as their partner is unknown
+                not request.env.user._is_public()
+                # Don't tokenize if the user tried to force it through the browser's developer tools
+                and acquirer_sudo.allow_tokenization
+                # Token is only created if required by the flow or requested by the user
+                and tokenization_required_or_requested
+            )
+        elif flow == 'token':  # Payment by token
+            token_sudo = request.env['payment.token'].sudo().browse(payment_option_id)
+            acquirer_sudo = token_sudo.acquirer_id
+            token_id = payment_option_id
+            tokenize = False
         else:
-            cid = user.company_id.id
-
-        # Check partner
-        if not user._is_public():
-            # NOTE: this means that if the partner was set in the GET param, it gets overwritten here
-            # This is something we want, since security rules are based on the partner - assuming the
-            # access_token checked out at the start, this should have no impact on the payment itself
-            # existing besides making reconciliation possibly more difficult (if the payment partner is
-            # not the same as the invoice partner, for example)
-            partner_id = user.partner_id.id
-        elif partner_id:
-            partner_id = int(partner_id)
-
-        values.update({
+            raise UserError(
+                _("The payment should either be direct, with redirection, or made by a token.")
+            )
+        reference = request.env['payment.transaction']._compute_reference(
+            acquirer_sudo.provider,
+            prefix=reference_prefix,
+            **(custom_create_values or {}),
+            **kwargs
+        )
+        if validation_route:  # Acquirers determine the amount and currency in validation operations
+            amount = acquirer_sudo._get_validation_amount()
+            currency_id = acquirer_sudo._get_validation_currency().id
+
+        # Create the transaction
+        tx_sudo = request.env['payment.transaction'].sudo().with_context(lang=None).create({
```

Comment:
why the `context=None`? it's only needed when you want to bypass the translation system (and usually for bad reasons)

Reply:
That was probably used before the apocalypse. I checked all three flows while removing the `with_context` in a debugger and it seems to work fine.